### PR TITLE
Clean up focus and context menu handling

### LINKED ISSUE #3556: Right-clicking on a cell input shows system menu when the cell is focused.
To reproduce (for me on Firefox 57 on OS X, but not Chrome):

1. Go into command mode
2. right-click on a cell input

effects:
1. The cell is changed to edit mode
2. the *system* context menu is shown, rather than the JLab notebook context menu.

Right-clicking on the cell input after the cell is already in edit mode correctly shows the JLab context menu.

Likely we're missing a `preventDefault` when switching the cell to edit mode? Or perhaps the Firefox 57 workaround in the code is somehow not finding the cell we're clicking on?

### LINKED ISSUE #3563: Multiple selection using CM editor
Question:

If someone is trying to do multiple selection, what should happen if they happen to click on the DM editor of the second cell that want to define the selection range?

In master, it will be treated as a regular click and focus the CM editor and put the cell into edit mode. Not sure if this is desired or not.

### PR DESCRIPTION
**Pull request by blink1073**
Fixes #3556. Fixes #3563.

### CODE REVIEW

**Comment by ellisonbg** [comment]
Still some awkward behavior in terms of what context menus pop up where:

Console editor area (seems ok):

screen shot 2018-01-06 at 11 32 19 am

Console input prompt area (the actual prompt area) has duplicates:

screen shot 2018-01-06 at 11 32 05 am

The CM editor of run cells shows the system native context menu (should show our context menu):

screen shot 2018-01-06 at 11 34 36 am

Other areas of previously run cells show this (fine):

screen shot 2018-01-06 at 11 35 27 am

Here is why I would aim for:

Everywhere:

* Restart Kernel
* Open Inspector

Only on Code cells:

* Clear Console Cells

**Comment by blink1073** [comment]
Note from video meeting: we are going to remove the `jp-mod-focus` class in the Notebook and add a note to extensions_dev about using a capture phase context menu listener to override our context menu.

File: tests/test-terminal/src/terminal.spec.ts [hunk 1]
```diff
@@ -255,7 +255,6 @@ describe('terminal/index', () => {
         expect(widget.methods).to.contain('onUpdateRequest');
         let style = window.getComputedStyle(widget.node);
         expect(style.backgroundColor).to.be('rgb(0, 0, 0)');
-        expect(style.color).to.be('rgb(255, 255, 255)');
```
Comment (ellisonbg) [inline tests/test-terminal/src/terminal.spec.ts]: Is this meant to be in this PR?

**Comment by ellisonbg** [comment]
Wonky bug:

* Create a MD cell followed by a code cell
* Render the markdown, and put it in active mode
* Right click on the Code cell below

Expected: the code cell should become active and its context menu should open

Observed: the markdown cell remains active, it is put into edit mode, and its context menu opens

jan-07-2018 08-15-31

Looks like hit testing is off - might be based on the markdown cells unrendered CM element rather than the rendered version?

**Comment by ellisonbg** [comment]
Recording the state of this branch (this is a known issue we will solve later):

When multiple cells are selected, right clicking on a cell's CM editor removes the multiple selection and puts the cell into edit mode.

**Comment by ellisonbg** [comment]
Overall this looks good, but there is the above bug related to the hit testing.

**Comment by blink1073** [comment]
I'm not able to reproduce the hit test bug.  Can we merge this and open that as a separate issue?

**Comment by blink1073** [comment]
Ah, it is happening in Chrome but not Firefox, looking...

**Comment by blink1073** [comment]
Okay, the issue is twofold: we are switching to edit mode, and then the input changes height so that the hit test takes us to a different cell.  If the markdown cell is the same size before and after rendering then the code cell is selected but the markdown cell is still in edit mode.

**Comment by ellisonbg** [comment]
I will watch this to see when it is ready for review...

**Comment by blink1073** [comment]
Done!

**Comment by ellisonbg** [comment]
testing locally...

1. Create a code and markdown cell.
2. Make sure the markdown cell is rendered
3. Put the code cell in edit mode.
4. Use the up/down arrows to navigate to the markdown cell -> the markdown cell will unrender and enter edit mode (this matches the classic notebook behavior)

Now, do the same thing (1-3) but *click* on the *rendered* markdown cell. In the classic notebook, the markdown cell will become active but remain rendered. In this branch, the markdown cell enters edit mode. I think we should stick to the classic notebook behavior in this respect.

Summary: clicking on a rendered markdown cell should select the cell, independent of what cell was active before and its state. Keyboard navigation with arrows should maintain edit mode (and thus unrender markdown cells).

**Comment by ellisonbg** [comment]
Ohh, and there is a follow on bug. When you click on a rendered markdown cell starting with a code cell in edit mode, you unrender the markdown cell and enter *neither* edit/command mode - the keyboard shortcuts stop working entirely for that cell.

**Comment by ellisonbg** [comment]
And more as well here is a animation:

jan-07-2018 15-00-18

**Comment by blink1073** [comment]
Tested on Firefox and Chrome, should be good to go.

- Left click outside of an editor always blurs the previous active cell
- Shift left click always expands selection
- Right click on an editor takes you to edit mode but does not affect the other previously selected cells.  Shift right click does the same but shows system menu
- Edit mode is preserved if the whole notebook loses focus (click on the file browser then on the notebook tab)

**Comment by blink1073** [comment]
Side note: `yarn watch --pattern src/widget.spec.ts` from the notebook tests directory was a delight to use.

**Comment by jasongrout** [comment]
On FF on OS X, in a notebook:

1. Create code cell
2. Put notebook in command mode
3. right-click on the code cell input codemirror. The notebook context menu correctly shows and the cell changes shading and border to indicate it is focused, but no cursor appears
4. Press escape to close the context menu. Now the CM looks like it is focused, but there is no cursor and you can't type in it:

screen shot 2018-01-08 at 8 49 53 am

**Comment by ian-r-rose** [comment]
I am seeing the same behavior as Jason, in both Firefox and Chrome.

**Comment by blink1073** [comment]
I made the changes discussed in the meeting, switching back to wip until I add some more tests for right mousedown behavior.

**Comment by jasongrout** [comment]
I just realized that this makes it impossible to split a cell with the right-click menu.

**Comment by jasongrout** [comment]
Basically, the command is disabled if the notebook is not in edit mode: https://github.com/jupyterlab/jupyterlab/blob/master/packages/notebook-extension/src/index.ts#L1053-L1059

I guess we'll need to just take out that entire check for now.

File: packages/notebook/src/widget.ts [hunk 2]
```diff
@@ -837,6 +843,13 @@ class Notebook extends StaticNotebook {
       newValue = Math.max(newValue, 0);
       newValue = Math.min(newValue, this.model.cells.length - 1);
     }
+
+    // Blur the previous active cell.
+    let activeCell = this.activeCell;
+    if (activeCell && newValue !== oldValue) {
+      activeCell.editor.blur();
```
Comment (jasongrout) [inline packages/notebook/src/widget.ts]: I think this should be merged into the check below, around line 857, where _activeCell is actually set to a new value.

File: packages/notebook/src/widget.ts [hunk 3]
```diff
@@ -1154,7 +1173,7 @@ class Notebook extends StaticNotebook {
   protected onAfterAttach(msg: Message): void {
     super.onAfterAttach(msg);
     let node = this.node;
-    node.addEventListener('mousedown', this);
+    node.addEventListener('mousedown', this, true);
```
Comment (jasongrout) [inline packages/notebook/src/widget.ts]: What is the advantage here of using a capture phase listener?

**Comment by ian-r-rose** [comment]
@jasongrout If we can split a cell when it is not in edit mode, there is no visual indication of where the split will happen, as the cursor is not showing. I am not sure if this is a better UX than not making `Split Cell` visible in the context menu.

**Comment by jasongrout** [comment]
> @jasongrout If we can split a cell when it is not in edit mode, there is no visual indication of where the split will happen, as the cursor is not showing. I am not sure if this is a better UX than not making Split Cell visible in the context menu.

1. We have the same issue in the main menu - it's not specific to the context menu.
2. Yes, I know it's not ideal. But with this patch, any way of invoking the split cell command other than the keyboard shortcut puts the cell into command mode.

File: packages/notebook/src/widget.ts [hunk 4]
```diff
@@ -865,6 +859,12 @@ class Notebook extends StaticNotebook {
     if (newValue === oldValue) {
       return;
     }
+
+    // Blur the previous active cell.
+    if (oldValue) {
+      oldValue.editor.blur();
+    }
+
```
Comment (jasongrout) [inline packages/notebook/src/widget.ts]: I was thinking something like:

```typescript
    if (cell !== this._activeCell) {
      // Post an update request.
      this.update();
      if (this._activeCell) {
        this._activeCell.editor.blur();
      }
      this._activeCell = cell;
      this._activeCellChanged.emit(cell);
    }
```

File: packages/notebook/src/widget.ts [hunk 5]
```diff
@@ -1354,16 +1366,29 @@ class Notebook extends StaticNotebook {
...
     // Shift right-click gives the browser default behavior.
-    if (event.shiftKey && event.button === 2) {
+    if (shiftKey && button === 2) {
       return;
     }
 
+    // Prevent a right click from temporarily focusing the cell and
+    // leaving the notebook in a quasi-state.
+    if (button === 2) {
```
Comment (jasongrout) [inline packages/notebook/src/widget.ts]: This is a pretty big sledgehammer - it says that *any* non-shift right-click event anywhere in the notebook is stopped in its tracks, right? This means that if a custom renderer has its own right-click handler, it's never called. Perhaps we should scope things a little more (e.g., if we're in a cell's input, and the target is in the editor?).

**Comment by jasongrout** [comment]
If we are in edit mode and press the down arrow key to go to the next cell, we now shift to command mode when we exit the cell, rather than go to the next cell in edit mode.

**Comment by jasongrout** [comment]
There are still a few issues, like the focus blinking off and on when clicking on a cell. I'm going to look at it more this morning.

**Comment by ian-r-rose** [comment]
I am also seeing the blue focus indicator blinking when right-clicking on the cell.

**Comment by jasongrout** [comment]
I'm looking into this - @blink1073, are you working on this right now?

File: packages/notebook/src/widget.ts [hunk 6]
```diff
@@ -1154,6 +1165,10 @@ class Notebook extends StaticNotebook {
   protected onAfterAttach(msg: Message): void {
     super.onAfterAttach(msg);
     let node = this.node;
+    // We must use to capture phase to prevent the code editor
+    // from receiving the mousedown so we can intercept right click
+    // event.
+    node.addEventListener('mousedown', this, true);
     node.addEventListener('mousedown', this);
```
Comment (ian-r-rose) [inline packages/notebook/src/widget.ts]: Is this double-mousdown-event-handling intended? It seems to be generating a lot of focus noise in my testing.

**Comment by jasongrout** [comment]
I have a commit or two to push that change some of the logic to make it clearer, I think.

File: packages/notebook/src/widget.ts (hunk 6 above)
Comment (jasongrout) [reply packages/notebook/src/widget.ts]: I have a commit or two to push that change some of the logic to make it clearer, I think.

**Comment by jasongrout** [comment]
I'm still trying to track down why a right click on an cell input area when in command mode flashes it to edit mode and then back to command mode, instead of just keeping it in command mode.

**Comment by jasongrout** [comment]
Discussion at https://gitter.im/jupyterlab/jupyterlab?at=5a555166ce68c3bc749bba7a about why we were seeing a flash in codemirror. Basically, in order to prevent codemirror from handling events, the event needs to have `preventDefault` set. So we needed to do that in the capture phase before the event reached codemirror. We were seeing a difference between browsers because on chrome, codemirror registers both mousedown and contextmenu handlers, while on firefox it only registers a mousedown, and handles the contextmenu event in the mousedown, IIRC.

In the long run, we need a way to prevent the editor from handling events that is editor-agnostic. Stopping propagation doesn't work as well, at least right now, since a lot of our logic is in our notebook handler in the bubble phase.

**Comment by ian-r-rose** [comment]
Tested locally, and it works much better now! Thanks @jasongrout for the detective work.

**Comment by jasongrout** [comment]
Thank you too for your detective work! And @blink1073!

**Comment by jasongrout** [comment]
I think this is probably good to go in for an rc2 - though I'd love to have some extra testing of this before releasing beta.

**Comment by blink1073** [comment]
Thanks!  Excelsior!

**Pull request merged**